dipy/viz/ui.py

```diff
@@ -0,0 +1,562 @@
+    - ui_list : This is used when there are more than one UI elements inside
+               a UI element. Inside the renderer, they're all iterated and added.
+    - parent_UI: This is useful of there is a parent UI element and its reference
+                needs to be passed down to the child.
+    """
+
+    def __init__(self):
+        self.ui_param = None
+        self.ui_list = list()
+
+        self.parent_UI = None
+        self._callbacks = []
+
+        self.left_button_state = "released"
+        self.right_button_state = "released"
+
+        self.handle_events()
+
+        self.on_left_mouse_button_pressed = lambda i_ren, obj, element: None
+        self.on_left_mouse_button_drag = lambda i_ren, obj, element: None
+        self.on_right_mouse_button_pressed = lambda i_ren, obj, element: None
+        self.on_right_mouse_button_drag = lambda i_ren, obj, element: None
+        self.on_mouse_hover = lambda i_ren, obj, element: None
+
+    def get_actors(self):
+        """ Returns the actors that compose this UI component. """
+        msg = "Subclasses of UI must implement `get_actors(self)`."
+        raise NotImplementedError(msg)
+
+    def add_to_renderer(self, ren):
+        """ Allows UI objects to add their own props to the renderer.
+        Parameters
+        ----------
+        ren : renderer
+        """
+        ren.add(*self.get_actors())
+
+        # Get a hold on the current interactor style.
+        iren = ren.GetRenderWindow().GetInteractor().GetInteractorStyle()
+
+        for callback in self._callbacks:
+            if not isinstance(iren, CustomInteractorStyle):
+                msg = ("The ShowManager requires `CustomInteractorStyle` in"
+                       " order to use callbacks.")
+                raise TypeError(msg)
+
+            iren.add_callback(*callback, args=[self])
+
+    def add_callback(self, prop, event_type, callback, priority=0):
+        """ Adds a callback to a specific event for this UI component.
+        Parameters
+        ----------
+        prop : vtkProp
+        event_type : event code
+        callback : function
+        priority : int
+        """
+        # Actually since we need an interactor style we will add the callback
+        # only when this UI component is added to the renderer.
+        self._callbacks.append((prop, event_type, callback, priority))
+
+    def set_center(self, position):
+        """ Sets the center of the UI component
+        Parameters
+        ----------
+        position : (float, float)
+        """
+        msg = "Subclasses of UI must implement `set_center(self, position)`."
+        raise NotImplementedError(msg)
+
+    def set_visibility(self, visibility):
+        """ Sets visibility of this UI component. """
+        for actor in self.get_actors():
+            actor.SetVisibility(visibility)
+
+    def handle_events(self):
+        self.add_callback("LeftButtonPressEvent", self.left_button_click_callback)
```

Comment:
Every subclass is supposed to have its own `add_callback`, which will have a `super.add_callback(prop, ....)`, where `prop` is the main element on which the events are to be added.